File: src-tauri/src/http/routes/transactions.rs
//! Sales: the list, one sale, and the act of making one.
//!
//! Everything here is in the `session` group, including the two operations only
//! an admin may perform — voiding a sale and correcting its payment method. The
//! router does not enforce that; `services::transactions::admin` does, with the
//! same `guard::require_admin` the Tauri path uses. Putting them in the admin
//! group as well would be belt and braces, but it would also put the route table
//! and the service at risk of disagreeing about who counts as an admin, and only
//! one of the two can be the answer.
//!
//! `POST /transactions` is the one route in the application with a header it
//! refuses to work without. See [`crate::http::idempotency`] for why.

use axum::body::Bytes;
use axum::extract::{Path, State};
use axum::http::{HeaderMap, StatusCode};
use axum::response::Response;
use axum::routing::{delete, get, patch, post};
use axum::{Extension, Router};
use chrono::Utc;

use crate::domain::receipt::ReceiptDataResponse;
use crate::domain::transactions::{
    CheckoutTransactionInput, DeleteTransactionInput, ListTransactionsInput, PaginatedTransactions,
    TransactionDetail, UpdatePaymentMethodInput,
};
use crate::domain::Actor;
use crate::entity::transactions;
use crate::http::error::ApiResult;
use crate::http::extract::{json_from_slice, Json, Query};
use crate::http::idempotency::{self, Claim};
use crate::http::AppState;
use crate::services;

pub fn session() -> Router<AppState> {
    Router::new()
        .route("/transactions", get(list).post(checkout))
        .route(
            "/transactions/next-receipt-number",
            get(next_receipt_number),
        )
        .route("/transactions/{id}", get(detail))
        .route("/transactions/{id}", delete(void))
        .route("/transactions/{id}/payment-method", patch(payment_method))
        .route("/transactions/{id}/receipt", get(receipt))
        .route("/transaction-items/{id}/ppob/retry", post(retry_ppob))
}

async fn list(
    State(state): State<AppState>,
    Query(input): Query<ListTransactionsInput>,
) -> ApiResult<axum::Json<PaginatedTransactions>> {
    Ok(axum::Json(
        services::transactions::list(&state.db, input).await?,
    ))
}

async fn next_receipt_number(State(state): State<AppState>) -> ApiResult<axum::Json<String>> {
    Ok(axum::Json(
        services::transactions::next_receipt_number(&state.db).await?,
    ))
}

async fn detail(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> ApiResult<axum::Json<TransactionDetail>> {
    Ok(axum::Json(
        services::transactions::detail(&state.db, id).await?,
    ))
}

/// Ring up a sale, at most once per `Idempotency-Key`.
///
/// The body is taken as bytes rather than through `Json<T>` because the key's
/// promise is about *this* request: the digest of what the client sent is what
/// distinguishes a retry from a different sale that happens to reuse a key.
///
/// The order is parse, then claim, then work. Parsing first means a malformed
/// body is rejected without consuming the key, so the client can fix it and
/// retry with the same one — which is what a client that generated one key per
/// cart will do.
async fn checkout(
    State(state): State<AppState>,
    Extension(actor): Extension<Actor>,
    headers: HeaderMap,
    body: Bytes,
) -> ApiResult<Response> {
    let key = idempotency::key_from_headers(&headers)?;
    let input: CheckoutTransactionInput = json_from_slice(&body)?;

    let guard = match idempotency::claim(
        &state.db,
        idempotency::SCOPE_CHECKOUT,
        actor.user_id,
        &key,
        &body,
        Utc::now(),
    )
    .await?
    {
        Claim::Replay(response) => return Ok(response),
        Claim::Fresh(guard) => guard,
    };

    match services::transactions::checkout(&state.db, &state.mitra, &actor, input).await {
        Ok(result) => Ok(guard.complete(&state.db, &result).await),
        Err(e) => {
            // A sale that did not happen must not block the corrected retry.
            guard.release(&state.db).await;
            Err(e.into())
        }
    }
}

#[derive(Debug, serde::Deserialize)]
struct VoidBody {
    reason: String,
}

/// Void a sale. Admin-only, enforced by the service.
async fn void(
    State(state): State<AppState>,
    Extension(actor): Extension<Actor>,
    Path(id): Path<i64>,
    Json(body): Json<VoidBody>,
) -> ApiResult<StatusCode> {
    services::transactions::void(
        &state.db,
        &actor,
        DeleteTransactionInput {
            transaction_id: id,
            reason: body.reason,
        },
    )
    .await?;
    Ok(StatusCode::NO_CONTENT)
}

#[derive(Debug, serde::Deserialize)]
struct PaymentMethodBody {
    payment_method: String,
    reason: String,
}

/// Correct the payment method on a completed sale. Admin-only, enforced by the
/// service, and audited by the `reason` it insists on.
async fn payment_method(
    State(state): State<AppState>,
    Extension(actor): Extension<Actor>,
    Path(id): Path<i64>,
    Json(body): Json<PaymentMethodBody>,
) -> ApiResult<axum::Json<transactions::Model>> {
    Ok(axum::Json(
        services::transactions::update_payment_method(
            &state.db,
            &actor,
            UpdatePaymentMethodInput {
                transaction_id: id,
                payment_method: body.payment_method,
                reason: body.reason,
            },
        )
        .await?,
    ))
}

/// Everything a receipt needs, for the browser to render or the printer route to
/// send.
async fn receipt(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> ApiResult<axum::Json<ReceiptDataResponse>> {
    Ok(axum::Json(
        services::receipt::receipt_data(&state.db, id).await?,
    ))
}

/// Re-run a PPOB line whose upstream fulfilment failed after the sale committed.
///
/// Not idempotency-guarded, and it does not need to be: the service claims the
/// item row before it calls upstream, so a second retry of an item already in
/// flight finds nothing to claim.
async fn retry_ppob(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> ApiResult<axum::Json<String>> {
    Ok(axum::Json(
        services::transactions::retry_ppob_fulfillment(&state.db, &state.mitra, id).await?,
    ))
}

File: src/lib/api/transactions.ts
import type { CheckoutTransactionInput, TransactionResult } from "@/features/cashier/types"
import type { ReceiptData, ReceiptLine } from "@/features/receipt/types"
import type {
  ListTransactionsInput,
  PaginatedTransactions,
  TransactionDetail,
} from "@/features/transactions/types"
import { apiDelete, apiGet, apiPatch, apiPost, type QueryParams } from "./client"

function listQuery(params: ListTransactionsInput): QueryParams {
  return {
    page: params.page,
    per_page: params.per_page,
    date_from: params.date_from,
    date_to: params.date_to,
    payment_method: params.payment_method,
    status: params.status,
    search: params.search,
  }
}

export function listTransactions(params: ListTransactionsInput): Promise<PaginatedTransactions> {
  return apiGet<PaginatedTransactions>("/transactions", listQuery(params))
}

export function getNextReceiptNumber(): Promise<string> {
  return apiGet<string>("/transactions/next-receipt-number")
}

export function getTransactionDetail(transactionId: number): Promise<TransactionDetail> {
  return apiGet<TransactionDetail>(`/transactions/${transactionId}`)
}

/**
 * Ring up a sale.
 *
 * `idempotencyKey` is not optional and the server rejects the request without
 * it. The key belongs to *the cart*, not to this function call: it is minted
 * once when a cart starts and reused by every attempt to check that cart out.
 * A retry after a lost response therefore replays the first sale instead of
 * ringing up a second one. See `useCartStore`, which owns the key's lifetime.
 */
export function checkoutTransaction(
  input: CheckoutTransactionInput,
  idempotencyKey: string,
): Promise<TransactionResult> {
  return apiPost<TransactionResult>("/transactions", input, { idempotencyKey })
}

/** Void a completed sale. Admin-only, enforced by the service. */
export function voidTransaction(transactionId: number, reason: string): Promise<void> {
  return apiDelete<void>(`/transactions/${transactionId}`, { reason })
}

export function updateTransactionPaymentMethod(
  transactionId: number,
  paymentMethod: string,
  reason: string,
): Promise<void> {
  return apiPatch<void>(`/transactions/${transactionId}/payment-method`, {
    payment_method: paymentMethod,
    reason,
  })
}

/** Everything a receipt needs, for the browser to render. */
export function getReceiptData(transactionId: number): Promise<ReceiptData> {
  return apiGet<ReceiptData>(`/transactions/${transactionId}/receipt`)
}

/**
 * The exact lines the printer would be handed for this sale, for the success
 * dialog's struk preview. `paperWidth` (58 or 80) overrides the store's
 * configured paper size; omitted, the server falls back to that same
 * setting — the one the print button itself uses, so the preview cannot
 * show a different width than what comes out of the printer.
 */
export function getSaleReceiptLines(
  transactionId: number,
  paperWidth?: number | null,
): Promise<ReceiptLine[]> {
  return apiGet<ReceiptLine[]>(`/transactions/${transactionId}/receipt/lines`, {
    paper: paperWidth ?? undefined,
  })
}

/**
 * Re-run a PPOB line whose upstream fulfilment failed after the sale committed.
 *
 * Only valid for an item in `failed`. An item in `processing` is already in
 * flight upstream and the server refuses to claim it twice. `pin` is the
 * cashier's Mitra transaction PIN, typed again for this attempt — the one
 * from the original sale was never kept anywhere to reuse.
 */
export function retryPpobFulfillment(transactionItemId: number, pin: string): Promise<string> {
  return apiPost<string>(`/transaction-items/${transactionItemId}/ppob/retry`, { pin })
}
